src/main/java/com/sequenceiq/cloudbreak/service/stack/flow/ProvisionUtil.java

```diff
@@ -36,24 +37,36 @@ public boolean isRequestFullWithCloudPlatform(Stack stack, int fullIndex) {
         return (fullIndex * instanceResourceBuilders.get(stack.cloudPlatform()).size()) % stack.cloudPlatform().parallelNumber() == 0;
     }
 
-    public void waitForRequestToFinish(Long stackId, List<Future<Boolean>> futures) {
+    public List<ResourceRequestResult> waitForRequestToFinish(Long stackId, List<Future<ResourceRequestResult>> futures)
+            throws ExecutionException, InterruptedException {
+        List<ResourceRequestResult> exceptions = new ArrayList<>();
         Stack stack = stackRepository.findOneWithLists(stackId);
         MDCBuilder.buildMdcContext(stack);
         LOGGER.info("Waiting for futures to finishing.");
-        StringBuilder sb = new StringBuilder();
-        Optional<Exception> exception = Optional.absent();
-        for (Future<Boolean> future : futures) {
-            try {
-                future.get();
-            } catch (Exception ex) {
-                exception = Optional.fromNullable(ex);
-                sb.append(String.format("%s, ", ex.getMessage()));
+        for (Future<ResourceRequestResult> future : futures) {
+            ResourceRequestResult resourceRequestResult = future.get();
+            if (FutureResult.FAILED.equals(resourceRequestResult.getFutureResult())) {
+                exceptions.add(resourceRequestResult);
             }
         }
-        if (exception.isPresent()) {
-            throw new BuildStackFailureException(sb.toString(), exception.orNull(), stack.getResources());
+        LOGGER.info("All futures finished continue with the next group.");
+        return exceptions;
+    }
+
+    public List<Resource> waitForRequestToFinishWithError(Long stackId, List<Future<ResourceRequestResult>> futures) throws Exception {
```

Comment:
Can we make 1 method from these two with a boolean throwException parameter? If it's true we throw an exception otherwise we'll return with a list of failed resources